### PR TITLE
ARROW-3158: [C++] Handle float truncation during casting

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/cast.h [hunk 1]
```diff
@@ -35,10 +35,23 @@ class DataType;
 namespace compute {
 
 struct ARROW_EXPORT CastOptions {
-  CastOptions() : allow_int_overflow(false), allow_time_truncate(false) {}
+  CastOptions()
+      : allow_int_overflow(false),
+        allow_time_truncate(false),
+        allow_float_truncate(true) {}
```
Comment (kszucs) [inline cpp/src/arrow/compute/kernels/cast.h]: This will be set to False by default via https://github.com/apache/arrow/pull/2504

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Comment by kszucs** [comment]
Thanks Uwe!

File: cpp/src/arrow/compute/kernels/cast.h [hunk 2]
```diff
@@ -35,10 +35,23 @@ class DataType;
...
+
+  explicit CastOptions(bool safe)
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/cast.h]: Perhaps annotate these methods as internal / experimental. I'm not sure what will be the best way longer-term to manage these flags

**Pull request closed without merging**